dart/optimizer/GradientDescentSolver.cpp

```diff
@@ -0,0 +1,482 @@
+void GradientDescentSolver::setMaxAttempts(size_t _maxAttempts)
+{
+  mGradientP.mMaxAttempts = _maxAttempts;
+}
+
+//==============================================================================
+size_t GradientDescentSolver::getMaxAttempts() const
+{
+  return mGradientP.mMaxAttempts;
+}
+
+//==============================================================================
+void GradientDescentSolver::setPerturbationStep(size_t _step)
+{
+  mGradientP.mPerturbationStep = _step;
+}
+
+//==============================================================================
+size_t GradientDescentSolver::getPerturbationStep() const
+{
+  return mGradientP.mPerturbationStep;
+}
+
+//==============================================================================
+void GradientDescentSolver::setMaxPerturbationFactor(double _factor)
+{
+  mGradientP.mMaxPerturbationFactor = _factor;
+}
+
+//==============================================================================
+double GradientDescentSolver::getMaxPerturbationFactor() const
+{
+  return mGradientP.mMaxPerturbationFactor;
+}
+
+//==============================================================================
+void GradientDescentSolver::setDefaultConstraintWeight(double _newDefault)
+{
+  mGradientP.mDefaultConstraintWeight = _newDefault;
+}
+
+//==============================================================================
+double GradientDescentSolver::getDefaultConstraintWeight() const
+{
+  return mGradientP.mDefaultConstraintWeight;
+}
+
+//==============================================================================
+Eigen::VectorXd& GradientDescentSolver::getEqConstraintWeights()
+{
+  return mGradientP.mEqConstraintWeights;
+}
+
+//==============================================================================
+const Eigen::VectorXd& GradientDescentSolver::getEqConstraintWeights() const
+{
+  return mGradientP.mEqConstraintWeights;
+}
+
+//==============================================================================
+Eigen::VectorXd& GradientDescentSolver::getIneqConstraintWeights()
+{
+  return mGradientP.mIneqConstraintWeights;
+}
+
+//==============================================================================
+const Eigen::VectorXd& GradientDescentSolver::getIneqConstraintWeights() const
+{
+  return mGradientP.mIneqConstraintWeights;
+}
+
+//==============================================================================
+void GradientDescentSolver::randomizeConfiguration(Eigen::VectorXd& _x)
+{
+  if(nullptr == mProperties.mProblem)
+    return;
+
+  if(_x.size() < static_cast<int>(mProperties.mProblem->getDimension()))
+    _x = Eigen::VectorXd::Zero(mProperties.mProblem->getDimension());
+
+  for(int i=0; i<_x.size(); ++i)
+  {
+    double lower = mProperties.mProblem->getLowerBounds()[i];
+    double upper = mProperties.mProblem->getUpperBounds()[i];
+    double step = upper - lower;
+    if(step > mGradientP.mMaxRandomizationStep)
+    {
+      step = 2*mGradientP.mMaxRandomizationStep;
```

Comment:
The naming of the variable is probably not great. ``step`` is really representing the difference between the upper bound and the lower bound of the number that will be randomly generated. When the difference between the upper and lower bound is too big, we clamp it down such that the original value for ``x`` is in the center of the range for the randomly generated value.

This condition statement makes it so that the lower bound will be ``x[i] - mMaxRandomizationStep`` and the upper bound will be ``x[i] + mMaxRandomizationStep``. A number that is randomly generated between those two bounds will be no further from ``x[i]`` than ``mMaxRandomizationStep``.

So I believe it is correct as it is, but could perhaps benefit from a better name.